docs/style-spec/expressions.md

```diff
@@ -0,0 +1,174 @@
+the appearance of a map feature to change with both the zoom level _and_ the
+feature's properties.
+
+```js
+{
+  "circle-radius": {
+    "expression": [
+      "curve", "linear", ["zoom"],
+
+      // zoom is 0 and "rating" is 0 -> circle radius will be 0px
+      // zoom is 0 and "rating" is 5 -> circle radius will be 5px
+      0, [ "get", ["properties"], "rating", "Number" ],
+      // zoom is 20 and "rating" is 0 -> circle radius will be 4 * 0 = 0px
+      // zoom is 20 and "rating" is 5 -> circle radius will be 4 * 5 = 20px
+      10, [ "*", 4, [ "get", ["properties"], "rating", "Number" ] ]
+    ]
+  }
+}
+```
+
+## Property Expressions
+
+Property expressions are represented using a Lisp-like structured syntax tree.
+
+###Types
+
+Every expression evaluates to a value of one of the following types.
+
+- `Null`
+  - Literal: `null`
+- `String`
+  - Literal: `"my string value"`
+- `Number`
+  - Literal: `1729`
+- `Boolean`
+  - Literal: `true` or `false`
+- `Color`
+- `JSONObject`
+- `Array<T, N>`: an array of N values of type T
+  - Literal: `["array", v0, v1, ...]`
+- `Vector<T>`: a dynamically sized array of values of type T
+  - Literal: `["vector", v0, v1, v2, ... vN]`
+  - TODO: without type inference, 0-length arrays and vectors can't be typed
+- `Value`: A "variant" type representing the set of possible values retrievable from a feature's `properties` object (`Null | String | Number | Boolean | JSONObject | Vector<Value>`)
+- `Error`: a subtype of all other types. Used wherever an expression is unable to return the appropriate supertype. Carries diagnostic information such as an explanatory message and the expression location where the Error value was generated.
+
+###Constants:
+- `[ "ln2" ] -> Number`
+- `[ "pi" ] -> Number`
+- `[ "e" ] -> Number`
+
+### Type conversion:
+- `["string", e: Value] -> String`
+- `["number", e:Value] -> Number`
+  - Uses platform-default string-to-number conversion. (TBD: parse locale-specificformatted number strings)
+- `["boolean", e:Value] -> Boolean`
+  - `0`, `''`, and `null` are converted to `false`, all other values to `true`
+
+###Lookup:
+- `["get", obj: JSONObject, key: String ] -> Value`
+- `["has", obj: JSONObject, key: String ] -> Boolean`
+- `["at", arr: JSONArray, index: Number] -> Value`
+- `["at", arr: Array<T>|Vector<T>, index: Number] -> T`
+- `["typeof", expr: Value] -> String`
+- `["length", e: JSONArray|Vector<T>|String] -> Number`
+
+###Feature data:
+- `["properties"] -> JSONObject` the feature's `properties` object
+- `["geometry_type"] -> String` the string value of `feature.geometry.type`
+- `[ "id" ] -> Value` returns the value of `feature.id`.
+
+###Decision:
+- `["case", cond1: Boolean, result1: T, cond2: Boolean, result2: T, ..., cond_m, result_m: T, result_otherwise: T] -> T`
+- `["match", x: T, a_1: T, y_1: U, a_2: T, y_2: U, ..., a_m: T, y_m: U, y_else: U]` - `a_1`, `a_2`, ... must be _literal_ values of type `T`.
```

Comment:
Can we generalize this so that multiple literal values can be mapped to the same output expression?